modules/apps/27-interchain-accounts/types/packet.go

```diff
@@ -47,6 +51,83 @@ func (iapd InterchainAccountPacketData) GetBytes() []byte {
 	return sdk.MustSortJSON(ModuleCdc.MustMarshalJSON(&iapd))
 }
 
+/**
+
+ADR-8 CallbackPacketData implementation
+
+InterchainAccountPacketData implements CallbackPacketDataI interface. This will allow middlewares targetting specific VMs
+to retrieve the desired callback addresses for the ICA packet on the source and destination chains.
+
+The Memo is used to set the desired callback addresses.
+
+The Memo format is defined like so:
+
+```json
+{
+	// ... other memo fields we don't care about
+	"callbacks": {
+		"src_callback_address": {contractAddrOnSrcChain},
+		"dest_callback_address": {contractAddrOnDestChain},
+		"src_callback_msg": {jsonObjectForSrcChainCallback},
+		"dest_callback_msg": {jsonObjectForDestChainCallback},
+	}
+
+}
+```
+
+**/
+
+// ADR-8 middleware should callback on the sender address on the source chain
+// if the sender address is an IBC Actor (i.e. smart contract that accepts IBC callbacks)
+func (iapd InterchainAccountPacketData) GetSrcCallbackAddress() (addr string) {
+	if len(iapd.Memo) == 0 {
+		return
+	}
+
+	jsonObject := make(map[string]interface{})
+	// the jsonObject must be a valid JSON object
+	err := json.Unmarshal([]byte(iapd.Memo), &jsonObject)
+	if err != nil {
+		return
+	}
+
+	callbackData, ok := jsonObject["callbacks"].(map[string]interface{})
+	if !ok {
+		return
+	}
+
+	callbackAddr := callbackData["src_callback_address"].(string)
+	return callbackAddr
+}
+
+// ADR-8 middleware should callback on the receiver address on the destination chain
+// if the receiver address is an IBC Actor (i.e. smart contract that accepts IBC callbacks)
+func (iapd InterchainAccountPacketData) GetDestCallbackAddress() (addr string) {
```

Comment:
I have the feeling this should just be disabled honestly. If you want custom logic to run on DEST, why not have it be a message encoded in the ICA tx itself?

I think for ICA, only sender callbacks make sense and for that I don't think we need a double confirmation in the memo

Reply:
I think it makes sense to disable this. I'm can't remember if ICA allows for multiple messages in one ICA tx but even if it doesn't, a callback would just extend this to two messages instead of multiple messages (which would be more useful)

Reply:
It does allow multiple messages